wallet: check when create wallets for the reserved name "wallets"

Saibato opened the pull request:
We allow since 0.17 to create a wallet with the the name or dirprefix :"wallets".

That might be not a good idea and we should discourage users to do that to prevent unexpected boating accidents and whale heart attacks.

edit@saibato
The wallets subdir is not created when chain == 'main' by default to allow the node to support upgrading or using older walletdirs or datadirs created by older core releases, 
testing could not catch this since this effects only mannet nodes, -regtest behave different and there we don't have this issue.

Saibato committed:
[wallet]: Skip creating wallets that include the default_walletdir name.

bugfix. wallets is no longer the safe word ;-)

laanwj commented:
Please add some explanation, why is "wallets" a special name?

Saibato commented:
> Please add some explanation, why is "wallets" a special name? Wallets are created, by default, in a directory named "wallets" but I don't see how creating a sub-directory in that named wallet should be problematic?
> 
> (and if so, isn't that a bug to be fixed and a functional test added for instead of worked around?)

Sorry, but afaics u can not test this with regtest,signet or testnet that effects only mainnet nodes.

zoltan-konder commented on src/wallet/wallet.cpp:
```diff
@@ -249,6 +249,13 @@ std::shared_ptr<CWallet> CreateWallet(interfaces::Chain& chain, const std::strin
 
+    // Sanity check bitcoin internal reserved
+    if (std::memcmp(name.c_str(), "wallets", 7) == 0) {
+        error = Untranslated("Please use a different walletname \"") + Untranslated(name) + Untranslated("\" - wallets is an reserved word.") + Untranslated(" ") + error;
```
```suggestion
        error = Untranslated("Please use a different walletname \"") + Untranslated(name) + Untranslated("\" - wallets is a reserved word.") + Untranslated(" ") + error;
```

zoltan-konder replied:
@Saibato can you fix it?

zoltan-konder reviewed (changes requested):
is **a** reserved word

Saibato commented:
@laanwj 
here a patch to test also mainnet nodes and this issue with our framework, that could be adapted to a full functional test :-)
 
```Diff
diff --git a/test/functional/test_framework/test_framework.py b/test/functional/test_framework/test_framework.py
index 4bda73599..440ff2672 100755
--- a/test/functional/test_framework/test_framework.py
+++ b/test/functional/test_framework/test_framework.py
@@ -403,7 +403,10 @@ class BitcoinTestFramework(metaclass=BitcoinTestMetaClass):
             n = self.nodes[i]
             if wallet_name is not None:
                 n.createwallet(wallet_name=wallet_name, descriptors=self.options.descriptors, load_on_startup=True)
-            n.importprivkey(privkey=n.get_deterministic_priv_key().key, label='coinbase')
+            if self.PRIV_KEYS is None:
+                n.importprivkey(privkey=n.get_deterministic_priv_key().key, label='coinbase')
+            else:
+                n.importprivkey(privkey=self.PRIV_KEYS[i].key, label='coinbase')
 
     def run_test(self):
         """Tests must override this method to define test logic"""
diff --git a/test/functional/test_framework/util.py b/test/functional/test_framework/util.py
index 123c48852..b5d334ffc 100644
--- a/test/functional/test_framework/util.py
+++ b/test/functional/test_framework/util.py
@@ -399,7 +399,10 @@ def get_auth_cookie(datadir, chain):
                     assert password is None  # Ensure that there is only one rpcpassword line
                     password = line.split("=")[1].strip("\n")
     try:
-        with open(os.path.join(datadir, chain, ".cookie"), 'r', encoding="ascii") as f:
+        sub = chain
+        if sub == "main" :
+            sub = ""
+        with open(os.path.join(datadir, sub, ".cookie"), 'r', encoding="ascii") as f:
             userpass = f.read()
             split_userpass = userpass.split(':')
             user = split_userpass[0]
@@ -413,7 +416,10 @@ def get_auth_cookie(datadir, chain):
 
 # If a cookie file exists in the given datadir, delete it.
 def delete_cookie_file(datadir, chain):
-    if os.path.isfile(os.path.join(datadir, chain, ".cookie")):
+    sub = chain
+    if chain == "main":
+        sub = ""
+    if os.path.isfile(os.path.join(datadir, sub, ".cookie")):
         logger.debug("Deleting leftover cookie file")
         os.remove(os.path.join(datadir, chain, ".cookie"))
 
diff --git a/test/functional/wallet_test.py b/test/functional/wallet_test.py
new file mode 100755
index 000000000..07d55fb83
--- /dev/null
+++ b/test/functional/wallet_test.py
@@ -0,0 +1,58 @@
+#!/usr/bin/env python3
+# Copyright (c) 2017-2019 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test wallet load on startup.
+
+Verify that a bitcoind node can maintain list of wallets loading on startup
+"""
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import (
+    assert_equal,
+    assert_raises_rpc_error,
+)
+import collections
+
+class WalletStartupTest(BitcoinTestFramework):
+
+    AddressKeyPair = collections.namedtuple('AddressKeyPair', ['address', 'key'])
+    MAIN_DET_PRIV_KEYS = [
+        # address , privkey
+        AddressKeyPair('bc1q4s7uj8djf97wp0z8ela90jz2wejfs5jy4e630t', 'KzNYJxz4SsNKpLeqtMcQ6gNfZv6GMdxAP6y8q8W5JtqUpBC2WWYw'),
+        AddressKeyPair('bc1q4adkrm53zvruavgsygnswtywrq04s8cslsnpjh', 'Kwc3y7M9P7Jewqmvh1XBqVhCUtMR4x49qGowxjDciFtB41aNHKyE'),
+    ]
+
+    def set_test_params(self):
+        self.chain = 'main'
+        self.num_nodes = 2
+        self.setup_clean_chain = True
+        self.PRIV_KEYS = self.MAIN_DET_PRIV_KEYS
+
+    def run_test(self):
+        self.log.info('test if we still hve the orginal wallets key')
+        self.nodes[0].createwallet("w1")
+        self.nodes[0].createwallet("w2")
+        assert_equal(self.nodes[0].listwallets(), ['', 'w1', 'w2'])
+        print(self.nodes[0].listwallets())
+        n = self.nodes[0].get_wallet_rpc('')
+        address = n.getnewaddress()
+        print(address)
+        print(n.dumpprivkey(address))
+        print(self.nodes[0].createwallet("wallets"))
+        n = self.nodes[0].get_wallet_rpc('')
+        print(n.dumpprivkey(address))
+        self.log.info('key still there')
+        print(address)
+        print(self.nodes[0].listwallets())
+        self.log.info('now restart node')
+        self.restart_node(0)
+        print(self.nodes[0].listwallets())
+        self.log.info('where is wallet w1 and w2 and where is the key from the default wallet?')
+        n = self.nodes[0].get_wallet_rpc('')
+        #oops
+        assert_raises_rpc_error(-4, " is not known", n.dumpprivkey, address)
+        self.log.info('oops')
+        #oops
+
+if __name__ == '__main__':
+    WalletStartupTest().main()

```

just patch and run
test/functional/wallet_test.py

MarcoFalke commented:
This should be unrelated to the chain. As long as there is a `wallet.dat` in the net specific datadir, and the user creates another wallet `wallets`, that folder will be treated as the new wallets dir.

Saibato commented:
> This should be unrelated to the chain

Should, yes, but...
So i think now about to split his into two PR.
A PR that enables by  placing the cookie in the correct path and adding address pairs for real fund coins easy mainnet testing, maybe even fund the real fund addresses on the fly from a dev fund with expensive test dust?

It seams to me clear that testing i.g. only confirms the test to run correct.
What happens if you use the real chain and funds could often be different and I wonder now what more can be found? hmm...

So this PR here is more a reminder fix and TODO.

Saibato committed:
Allow self.chain = 'main' in python bitcoin test framework

mainchain and testchains could behave different.
So make sure we can now easy test this.

Saibato commented:
@laanwj 

> (and if so, isn't that a bug to be fixed and a functional test added for instead of worked around?)

I thought about that and but if we want to support seamless upgrede and use of older releases i see not much other way, considering that this effects also new sqlite and adaptor wallets from newer releases, it seams to me the right way to just skip creating a wallets root subdir by chance.
